File: workload_data_generation/test_fre_sort.cpp
#include <fstream>
#include <iostream>
#include <sstream>
#include <map>
#include <vector>
#include <algorithm>
#include <string>

// 定义用于存储键和频率的pair的类型
typedef std::pair<int64_t, int> KeyFreqPair;

// 定义比较函数，用于排序
bool compareFreq(const KeyFreqPair& a, const KeyFreqPair& b) {
    return a.second > b.second; // 降序排序
}

int main() {

    std::string inputFileName = "/home/jeff-wang/workloads/zipf1.02_keys10.0B.csv";
    std::string outputFileName = "/home/jeff-wang/workloads/cpp_test.csv";
    std::map<int64_t, int> frequencies;

    std::ifstream inputFile(inputFileName);
    std::ofstream outputFile(outputFileName);

    if (!inputFile.is_open() || !outputFile.is_open()) {
        std::cerr << "Error opening file." << std::endl;
        return 1;
    }

    std::string line;
    // Skip header
    std::getline(inputFile, line);

    while (std::getline(inputFile, line)) {
        std::stringstream ss(line);
        std::string keyString;
        std::getline(ss, keyString, ',');

        int64_t key = std::stoll(keyString);
        frequencies[key]++;
    }

    // 将map的内容复制到vector中以便排序
    std::vector<KeyFreqPair> freqVector(frequencies.begin(), frequencies.end());

    // 使用自定义比较函数对vector进行排序
    std::sort(freqVector.begin(), freqVector.end(), compareFreq);

    // 写入到输出文件
    outputFile << "Key,Frequency\n";
    for (const auto& pair : freqVector) {
        outputFile << pair.first << "," << pair.second << "\n";
    }

    std::cout << "Frequencies have been written to " << outputFileName << std::endl;

    return 0;
}

File: workload_data_generation/test_fre_sort.py
import csv
from collections import defaultdict

def calculate_key_frequencies(input_file_name, output_file_name):
    # 使用defaultdict来计数，初始化为0
    frequencies = defaultdict(int)

    # 从CSV文件中读取数据
    with open(input_file_name, mode='r', newline='') as file:
        reader = csv.reader(file)
        header = next(reader)  # 跳过标题行
        for row in reader:
            key = row[0]  # 假设key位于每行的第一列
            frequencies[key] += 1

    frequencies_sorted = sorted(frequencies.items(), key=lambda item: item[1], reverse=True)

    # 将频率结果写入到一个新的CSV文件中
    with open(output_file_name, mode='w', newline='') as file:
        writer = csv.writer(file)
        writer.writerow(['Key', 'Frequency'])  # 写入标题行
        for key, freq in frequencies_sorted:
            writer.writerow([key, freq])

# 示例使用
input_file_name = "/home/jeff-wang/workloads/zipf1.02_keys10.0B.csv"
output_file_name = "/home/jeff-wang/workloads/python_test.csv"
calculate_key_frequencies(input_file_name, output_file_name)
